Sherlock/replicator.h

```diff
@@ -0,0 +1,296 @@
+        const auto remote_schema = ParseJSON<SubscribableSherlockSchema>(response.body);
+        if (remote_schema != schema_) {
+          CURRENT_THROW(RemoteStreamInvalidSchemaException());
+        }
+      } else {
+        CURRENT_THROW(RemoteStreamDoesNotRespondException());
+      }
+    }
+
+    const std::string GetURLToSubscribe(uint64_t index) const {
+      return url_ + "?i=" + current::ToString(index);
+    }
+
+    const std::string GetURLToTerminate(const std::string& subscription_id) {
+      return url_ + "?terminate=" + subscription_id;
+    }
+
+   private:
+    const std::string url_;
+    const SubscribableSherlockSchema schema_;
+  };
+
+  template <typename F, typename TYPE_SUBSCRIBED_TO>
+  class RemoteSubscriberThread final : public current::sherlock::SubscriberScope::SubscriberThread {
+   public:
+    RemoteSubscriberThread(ScopeOwned<RemoteStream>& remote_stream,
+                           F& subscriber,
+                           uint64_t start_idx,
+                           std::function<void()> done_callback)
+        : valid_(false),
+          done_callback_(done_callback),
+          external_stop_(false),
+          remote_stream_(remote_stream, [this]() { TerminateSubscription(); }),
+          subscriber_(subscriber),
+          index_(start_idx),
+          thread_([this]() { Thread(); }) {
+      valid_ = true;
+    }
+
+    ~RemoteSubscriberThread() {
+      if (valid_) {
+        if (!subscriber_thread_done_) {
+          TerminateSubscription();
+        }
+        thread_.join();
+      } else {
+        if (done_callback_) {
+          done_callback_();
+        }
+      }
+    }
+
+   private:
+    static_assert(current::ss::IsEntrySubscriber<F, TYPE_SUBSCRIBED_TO>::value, "");
+    RemoteSubscriberThread() = delete;
+    RemoteSubscriberThread(const RemoteSubscriberThread&) = delete;
+    RemoteSubscriberThread(RemoteSubscriberThread&&) = delete;
+    void operator=(const RemoteSubscriberThread&) = delete;
+    void operator=(RemoteSubscriberThread&&) = delete;
+
+    void Thread() {
+      ThreadImpl();
+      subscriber_thread_done_ = true;
+      if (done_callback_) {
+        done_callback_();
+      }
+    }
+
+    void ThreadImpl() {
+      const RemoteStream& bare_stream = remote_stream_.ObjectAccessorDespitePossiblyDestructing();
+      bool terminate_sent = false;
+      while (true) {
+        if (!terminate_sent && external_stop_) {
+          terminate_sent = true;
+          if (subscriber_.Terminate() != ss::TerminationResponse::Wait) {
+            return;
+          }
+        }
+        try {
+          bare_stream.CheckSchema();
+          HTTP(ChunkedGET(
+              bare_stream.GetURLToSubscribe(index_),
+              [this](const std::string& header, const std::string& value) { OnHeader(header, value); },
+              [this](const std::string& chunk_body) { OnChunk(chunk_body); },
+              [this]() {}));
+        } catch (StreamTerminatedBySubscriber&) {
```

Comment:
`OnHeader()` here would save `subscription_id_`. Won't it make sense to erase it after `HTTP(ChunkedGET(...))` returns, as its value is no longer valid? (Or perhaps combine `subscription_id_` and `subscription_established_` in one `WaitableAtomic<>` would be a good idea?)